File: frontend/pages/concordancia.go
package pages

import "strings"

// Spanish gender/number agreement (concordancia) helpers for the HTML draft
// previews — a Go port of pythonapi/concordancia.py so the on-screen preview
// matches the generated .docx. Every function operates on the full party
// list, however long (never assumes 1 or 2 people).

const (
	GenderM = "masculin"
	GenderF = "femenin"
)

// agree picks the form matching gender; falls back to the ambiguous "(a)"
// hack (matching the templates' pre-existing convention) when unknown.
func agree(masc, fem, gender string) string {
	switch gender {
	case GenderF:
		return fem
	case GenderM:
		return masc
	default:
		return masc + "(a)"
	}
}

var maritalStatusES = map[string][2]string{
	"single":      {"soltero", "soltera"},
	"married":     {"casado", "casada"},
	"divorced":    {"divorciado", "divorciada"},
	"widowed":     {"viudo", "viuda"},
	"legal_union": {"en unión libre", "en unión libre"},
}

// maritalStatusEs translates the DB's English marital-status code to
// gendered Spanish. Text that isn't one of the known codes (e.g. already
// hand-typed Spanish via the placeholder editor) passes through unchanged.
func maritalStatusEs(raw, gender string) string {
	if raw == "" {
		return ""
	}
	pair, ok := maritalStatusES[strings.ToLower(strings.TrimSpace(raw))]
	if !ok {
		return raw
	}
	switch gender {
	case GenderF:
		return pair[1]
	case GenderM:
		return pair[0]
	default:
		if pair[0] == pair[1] {
			return pair[0]
		}
		return pair[0] + "(a)"
	}
}

func quienWord(n int) string {
	if n == 1 {
		return "quien"
	}
	return "quienes"
}

func sigueWord(n int) string {
	if n == 1 {
		return "sigue"
	}
	return "siguen"
}

func denominarWord(n int) string {
	if n == 1 {
		return "se denominará"
	}
	return "se denominarán"
}

func haWord(n int) string {
	if n == 1 {
		return "ha"
	}
	return "han"
}

// pickWord is a generic singular/plural picker for one-off verbs tied to a
// party count (e.g. "acepta"/"aceptan", "da"/"dan").
func pickWord(n int, singular, plural string) string {
	if n == 1 {
		return singular
	}
	return plural
}

// composition returns "" if no one's gender is known, GenderF only if every
// person with a known gender is a woman, GenderM otherwise (mixed, all men,
// or an unknown gender alongside at least one man).
func composition(parties []PartyPreview) string {
	sawF, sawM := false, false
	for _, p := range parties {
		switch p.Gender {
		case GenderF:
			sawF = true
		case GenderM:
			sawM = true
		}
	}
	if !sawF && !sawM {
		return ""
	}
	if sawF && !sawM {
		return GenderF
	}
	return GenderM
}

// (masculine singular, feminine singular, masculine plural, feminine plural)
// — the 4 forms of each role denomination, matching today's fixed defaults
// in builder.templ (e.g. denomDefault: 'LOS VENDEDORES').
var denominations = map[string][4]string{
	"VENDEDOR":      {"EL VENDEDOR", "LA VENDEDORA", "LOS VENDEDORES", "LAS VENDEDORAS"},
	"COMPRADOR":     {"EL COMPRADOR", "LA COMPRADORA", "LOS COMPRADORES", "LAS COMPRADORAS"},
	"PROPIETARIO":   {"EL PROPIETARIO", "LA PROPIETARIA", "LOS PROPIETARIOS", "LAS PROPIETARIAS"},
	"INQUILINO":     {"EL INQUILINO", "LA INQUILINA", "LOS INQUILINOS", "LAS INQUILINAS"},
	"PODERDANTE":    {"EL PODERDANTE", "LA PODERDANTE", "LOS PODERDANTES", "LAS PODERDANTES"},
	"APODERADO":     {"EL APODERADO", "LA APODERADA", "LOS APODERADOS", "LAS APODERADAS"},
	"COMPARECIENTE": {"EL COMPARECIENTE", "LA COMPARECIENTE", "LOS COMPARECIENTES", "LAS COMPARECIENTES"},
	"DECLARANTE":    {"EL DECLARANTE", "LA DECLARANTE", "LOS DECLARANTES", "LAS DECLARANTES"},
	"AUTORIZANTE":   {"EL AUTORIZANTE", "LA AUTORIZANTE", "LOS AUTORIZANTES", "LAS AUTORIZANTES"},
	"ACEPTANTE":     {"EL ACEPTANTE", "LA ACEPTANTE", "LOS ACEPTANTES", "LAS ACEPTANTES"},
}

// ResolveDenomination recomputes `raw` in the correct gender/number for
// `parties` if it is still one of the role's known defaults; if the caller
// customized it to other text, it is left untouched.
func ResolveDenomination(raw, role string, parties []PartyPreview) string {
	forms, ok := denominations[role]
	if !ok || len(parties) == 0 {
		return raw
	}
	upper := strings.ToUpper(strings.TrimSpace(raw))
	matches := false
	for _, f := range forms {
		if f == upper {
			matches = true
			break
		}
	}
	if !matches {
		return raw
	}
	comp := composition(parties)
	if len(parties) == 1 {
		switch comp {
		case GenderF:
			return forms[1]
		case GenderM:
			return forms[0]
		default:
			return raw // unknown gender, single person: don't guess
		}
	}
	if comp == GenderF {
		return forms[3]
	}
	return forms[2]
}

File: pythonapi/concordancia.py
"""Spanish gender/number agreement (concordancia) helpers for party
descriptions in notarial acts and contracts.

Every party role (sellers, buyers, declarants, witnesses, ...) is a list of
people, never a fixed count of 1 or 2 — every function here operates on the
full list, however long.
"""

from docxtpl import RichText

GENDER_M = "masculin"
GENDER_F = "femenin"


def agree(masc: str, fem: str, gender: str | None) -> str:
    """Pick the form matching `gender`; falls back to the ambiguous "(a)"
    hack (matching the templates' pre-existing convention) when unknown."""
    if gender == GENDER_F:
        return fem
    if gender == GENDER_M:
        return masc
    return f"{masc}(a)"


MARITAL_STATUS_ES: dict[str, tuple[str, str]] = {
    "single": ("soltero", "soltera"),
    "married": ("casado", "casada"),
    "divorced": ("divorciado", "divorciada"),
    "widowed": ("viudo", "viuda"),
    "legal_union": ("en unión libre", "en unión libre"),
}


def marital_status_es(raw: str | None, gender: str | None) -> str:
    """Translate the DB's English marital-status code to gendered Spanish.
    Text that isn't one of the known codes (e.g. already hand-typed Spanish
    via the placeholder editor) passes through unchanged."""
    if not raw:
        return ""
    pair = MARITAL_STATUS_ES.get(raw.strip().lower())
    if pair is None:
        return raw
    masc, fem = pair
    if gender == GENDER_F:
        return fem
    if gender == GENDER_M:
        return masc
    return masc if masc == fem else f"{masc}(a)"


def describe_party(p) -> str:
    """One person's descriptive clause, gender-agreeing wherever their
    gender is known. `p` is a CustomerInfo/HeirInfo-like object exposing
    nationality, marital_status, occupation, id_number, address, gender."""
    parts: list[str] = []
    if p.nationality:
        # Nationality agrees with the feminine noun "nacionalidad", not with
        # the person — stays feminine regardless of gender (see
        # themisdatacrawler ResolveNationality for the same convention).
        parts.append(f"de nacionalidad {p.nationality}")
    parts.append("mayor de edad")
    ms = marital_status_es(getattr(p, "marital_status", None), p.gender)
    if ms:
        parts.append(ms)
    if p.occupation:
        parts.append(p.occupation)
    if p.id_number:
        portador = agree("portador", "portadora", p.gender)
        parts.append(f"{portador} de la cédula de identidad y electoral No. {p.id_number}")
    if p.address:
        domiciliado = agree("domiciliado", "domiciliada", p.gender)
        parts.append(f"{domiciliado} y residente en {p.address}")
    return ", ".join(parts)


def party_block(people) -> RichText:
    """Bold-upper name + gendered description per person, separated by
    '; ' — the direct replacement for the old shared-adjective sentence."""
    rt = RichText()
    for i, p in enumerate(people):
        if i:
            rt.add("; ")
        rt.add(f"{p.first_name} {p.last_name}".upper(), bold=True)
        rt.add(f", {describe_party(p)}")
    return rt


def quien(n: int) -> str:
    return "quien" if n == 1 else "quienes"


def sigue(n: int) -> str:
    return "sigue" if n == 1 else "siguen"


def denominar(n: int) -> str:
    return "se denominará" if n == 1 else "se denominarán"


def ha(n: int) -> str:
    return "ha" if n == 1 else "han"


def pick(n: int, singular: str, plural: str) -> str:
    """Generic singular/plural picker for one-off verbs tied to a party
    count (e.g. "acepta"/"aceptan", "da"/"dan")."""
    return singular if n == 1 else plural


def _composition(people) -> str | None:
    """None if no one's gender is known; GENDER_F only if every person with
    a known gender is a woman; GENDER_M otherwise (mixed, all men, or an
    unknown gender alongside at least one man)."""
    known = {p.gender for p in people if p.gender in (GENDER_M, GENDER_F)}
    if not known:
        return None
    return GENDER_F if known == {GENDER_F} else GENDER_M


# (masculine singular, feminine singular, masculine plural, feminine plural)
# — the 4 forms of each role denomination, matching today's fixed defaults
# in builder.templ (e.g. denomDefault: 'LOS VENDEDORES').
DENOMINATIONS: dict[str, tuple[str, str, str, str]] = {
    "VENDEDOR": ("EL VENDEDOR", "LA VENDEDORA", "LOS VENDEDORES", "LAS VENDEDORAS"),
    "COMPRADOR": ("EL COMPRADOR", "LA COMPRADORA", "LOS COMPRADORES", "LAS COMPRADORAS"),
    "PROPIETARIO": ("EL PROPIETARIO", "LA PROPIETARIA", "LOS PROPIETARIOS", "LAS PROPIETARIAS"),
    "INQUILINO": ("EL INQUILINO", "LA INQUILINA", "LOS INQUILINOS", "LAS INQUILINAS"),
    "PODERDANTE": ("EL PODERDANTE", "LA PODERDANTE", "LOS PODERDANTES", "LAS PODERDANTES"),
    "APODERADO": ("EL APODERADO", "LA APODERADA", "LOS APODERADOS", "LAS APODERADAS"),
    "COMPARECIENTE": ("EL COMPARECIENTE", "LA COMPARECIENTE", "LOS COMPARECIENTES", "LAS COMPARECIENTES"),
    "DECLARANTE": ("EL DECLARANTE", "LA DECLARANTE", "LOS DECLARANTES", "LAS DECLARANTES"),
    "AUTORIZANTE": ("EL AUTORIZANTE", "LA AUTORIZANTE", "LOS AUTORIZANTES", "LAS AUTORIZANTES"),
    "ACEPTANTE": ("EL ACEPTANTE", "LA ACEPTANTE", "LOS ACEPTANTES", "LAS ACEPTANTES"),
}


def resolve_denomination(raw: str, role: str, people) -> str:
    """If `raw` is still one of the role's known defaults, recompute it in
    the correct gender/number for `people`; if the caller customized it to
    other text, leave it untouched."""
    forms = DENOMINATIONS.get(role)
    if not people or forms is None or raw.strip().upper() not in forms:
        return raw
    masc_s, fem_s, masc_p, fem_p = forms
    comp = _composition(people)
    if len(people) == 1:
        if comp == GENDER_F:
            return fem_s
        if comp == GENDER_M:
            return masc_s
        return raw  # unknown gender, single person: don't guess
    return fem_p if comp == GENDER_F else masc_p
